### ISSUE TITLE
[1.19.3] [RFC] Creative mode tab content default ordering

### ISSUE THREAD

**Issue by Minecraftschurli** [MinecraftForge/MinecraftForge#9192]
This RFC is to discuss the default ordering of item stacks in creative mode tabs (or the lack thereof).
Currently there is absolutely no guaranty to ordering if one doesn't specify it explicitly.
My proposal is that the default ordering is insertion order (the order of calls to accept) so people don't have to think about before and after and can just call accept in the order they want.

**Comment by marchermans** [comment]
> This RFC is to discuss the default ordering of item stacks in creative mode tabs (or the lack thereof). Currently there is absolutely no guaranty to ordering if one doesn't specify it explicitly. My proposal is that the default ordering is insertion order (the order of calls to accept) so people don't have to think about before and after and can just call accept in the order they want.

So when I designed this initially there is the effect of the registration order of the items yes, but also the effect of the event handling, which happens on the mod bus, so there is no direct relationship available.

Which is why the populator has methods which accept sorting order.

Not sure what we should do with the items that are added without sorting order though?
Always last?

**Comment by Raycoms** [comment]
I agree with always last

**Comment by TelepathicGrunt** [comment]
My vote is always last too. Basically if I add items like this, or with a list of items, the items stay in the same order as define here

```java
public static void registerCreativeModeTab(CreativeModeTabEvent.Register event) {
        BUMBLEZONE_TAB = event.registerCreativeModeTab(new ResourceLocation(Bumblezone.MODID, "main_tab"),
                builder -> builder.icon(() -> new ItemStack(BzItems.HONEYCOMB_BROOD.get()))
                .title(Component.translatable("itemGroup." + Bumblezone.MODID + ".main_tab"))
                .withLabelColor(0xDDBB00)
                .displayItems((features, output, hasPermissions) -> {
                    output.accept(new ItemStack(BzItems.POROUS_HONEYCOMB.get()));
                    output.accept(new ItemStack(BzItems.FILLED_POROUS_HONEYCOMB.get()));
                    output.accept(new ItemStack(BzItems.EMPTY_HONEYCOMB_BROOD.get()));
                    output.accept(new ItemStack(BzItems.HONEYCOMB_BROOD.get()));
                    output.accept(new ItemStack(BzItems.STICKY_HONEY_RESIDUE.get()));
                    output.accept(new ItemStack(BzItems.STICKY_HONEY_REDSTONE.get()));
               }));
}
```

**Comment by marchermans** [comment]
> My vote is always last too. Basically if I add items like this, or with a list of items, the items stay in the same order as define here
> 
> ```java
> public static void registerCreativeModeTab(CreativeModeTabEvent.Register event) {
>         BUMBLEZONE_TAB = event.registerCreativeModeTab(new ResourceLocation(Bumblezone.MODID, "main_tab"),
>                 builder -> builder.icon(() -> new ItemStack(BzItems.HONEYCOMB_BROOD.get()))
>                 .title(Component.translatable("itemGroup." + Bumblezone.MODID + ".main_tab"))
>                 .withLabelColor(0xDDBB00)
>                 .displayItems((features, output, hasPermissions) -> {
>                     output.accept(new ItemStack(BzItems.POROUS_HONEYCOMB.get()));
>                     output.accept(new ItemStack(BzItems.FILLED_POROUS_HONEYCOMB.get()));
>                     output.accept(new ItemStack(BzItems.EMPTY_HONEYCOMB_BROOD.get()));
>                     output.accept(new ItemStack(BzItems.HONEYCOMB_BROOD.get()));
>                     output.accept(new ItemStack(BzItems.STICKY_HONEY_RESIDUE.get()));
>                     output.accept(new ItemStack(BzItems.STICKY_HONEY_REDSTONE.get()));
>                }));
> }
> ```

That order is kept, since that is part of the original tabs builder.
Well it is guaranteed that they appear in that order, but not that no other mod can not inject other items in between.

**Comment by Minecraftschurli** [comment]
But if you don't do it on your own tab but on a vanilla one then nothing is guaranteed if you don't specify the before/after.

**Comment by shedaniel** [comment]
What about if you are trying to add a list of entries to be behind one vanilla item?

e.g. Adding different colored variants of shulker boxes?
In that case do you need to declare order relationships between the different colored variants?

IMO, that's annoying. Calling `populator.acceptAll(Collection<ItemStack>)` should at least ensure that the stacks within the collection is ordered from each other.

**Comment by LexManos** [comment]
Closed via #9198
